mbf_abstract_nav/include/mbf_abstract_nav/abstract_action.h

```diff
@@ -51,106 +50,109 @@ template <typename Action, typename Execution>
-    if(slot != concurrency_slots_.left.end()) // if there is a plugin running on the same slot, cancel it // TODO make thread safe
+    uint8_t slot = goal_handle.getGoal()->concurrency_slot;
+
+    boost::lock_guard<boost::mutex> guard(slots_mtx_);
+    typename std::map<uint8_t, ConcurrencySlot>::iterator slot_it = concurrency_slots_.find(slot);
+    if(slot_it != concurrency_slots_.end())
     {
-      typename std::map<const std::string, const typename Execution::Ptr>::const_iterator elem
-          = executions_.find(slot->second);
-      if(elem != executions_.end())
-      {
-        elem->second->cancel();
-      }
-      concurrency_slots_.left.erase(slot->first);
+      // if there is a plugin running on the same slot, cancel it
+      slot_it->second.execution->cancel(); // TODO make thread safe
     }
-    concurrency_slots_.insert(SlotGoalIdMap::value_type(goal_handle.getGoal()->concurrency_slot, goal_handle.getGoalID().id));
-    executions_.insert(std::pair<const std::string, const typename Execution::Ptr>(goal_handle.getGoalID().id, execution_ptr));
-    threads_ptrs_.insert(std::pair<const std::string, boost::thread*>(goal_handle.getGoalID().id,
-                         threads_.create_thread(boost::bind(&AbstractAction::runAndCleanUp, this, goal_handle, execution_ptr))));
+    // fill concurrency slot with the new goal handle, execution, and working thread
+    concurrency_slots_[slot].goal_handle = goal_handle;
+    concurrency_slots_[slot].execution = execution_ptr;
+    concurrency_slots_[slot].thread_ptr = threads_.create_thread(boost::bind(&AbstractAction::runAndCleanUp,
+                                                                                 this, goal_handle, execution_ptr));
   }
 
-  void cancel(GoalHandle &goal_handle){
-    typename std::map<const std::string, const typename Execution::Ptr>::const_iterator
-        elem = executions_.find(goal_handle.getGoalID().id);
-    if(elem != executions_.end())
+  virtual void cancel(GoalHandle &goal_handle){
+    uint8_t slot = goal_handle.getGoal()->concurrency_slot;
+
+    boost::lock_guard<boost::mutex> guard(slots_mtx_);
+    typename std::map<uint8_t, ConcurrencySlot>::iterator slot_it = concurrency_slots_.find(slot);
+    if(slot_it != concurrency_slots_.end())
     {
-      boost::lock_guard<boost::mutex> guard(map_mtx_);
-      elem->second->cancel();
+      concurrency_slots_[slot].execution->cancel();
     }
   }
 
-  void runAndCleanUp(GoalHandle goal_handle, typename Execution::Ptr execution_ptr){
+  virtual void runAndCleanUp(GoalHandle &goal_handle, typename Execution::Ptr execution_ptr){
+    uint8_t slot = goal_handle.getGoal()->concurrency_slot;
+
     if (execution_ptr->setup_fn_)
       execution_ptr->setup_fn_();
-    run_(goal_handle, *execution_ptr);
-    ROS_DEBUG_STREAM("Finished action run method, waiting for execution thread to finish.");
+    run_(concurrency_slots_[slot].goal_handle, *execution_ptr);
```

Comment:
Aa here you use the map version of the `goal_handle`. Shouldn't it then already work with the reference version and not the map version in the run method?

Reply:
See my comment above. Unless we switch to pointers everywhere, we must stick to the map version (uglier, I know :see_no_evil: )

Reply:
I know the difference between ptr and ref. But see the documentation of the goal handle. The object stays the same. Only the internal fields are changed using the assign operator

Reply:
mmm...  you are right...  but I tried the change you propose and it stopped working! So I assume the run method keeps accessing the old GoalHandle. I need to investigate what's really going on here

Reply:
I have tracked the goal id and I can confirm that the overlaoded assignment operator you mentioned doesn't affect already created references. Then, we have only two options: let the PR as it is now, or change all to pointers

Reply:
Maybe you need to define a ref in the struct and hand over the ref via a constructor. Could you try that as the last choice.